### PR TITLE
Tokio compatibility

### PR DESCRIPTION
**Pull request by Nemo157**
I have tested the example in an external project, I'm not sure what to do about having it actually tested since you can't have optional dev-dependencies and I really don't want to bring in all of `tokio` normally.

### CODE REVIEW

File: futures-util/src/compat/tokio.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+};
+use tokio_executor::{DefaultExecutor, Executor as _};
+
+/// An executor that delegates to `tokio`'s [`DefaultExecutor`], will panic if
+/// used in the context of a task that is not running on `tokio`'s executor.
+///
+/// # Examples
+///
+/// ```ignore
+/// #![feature(async_await, await_macro, underscore_imports, pin)]
```
Comment (MajorBreakfast) [inline futures-util/src/compat/tokio.rs]: Is the `underscore_imports` feature required?

File: futures-util/src/compat/tokio.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+///     future::TryFutureExt as _,
+///     executor::spawn,
+///     compat::TokioDefaultExecutor,
+/// };
+///
+/// let (sender, receiver) = oneshot::channel::<i32>();
+///
+/// let handle = thread::spawn(move || {
+///     futures::executor::block_on(async move {
+///         await!(receiver).unwrap()
```
Comment (MajorBreakfast) [inline futures-util/src/compat/tokio.rs]: Why not `block_on(receiver).unwrap()` at the bottom? Does it have to be in another thread?

File: futures-util/src/compat/tokio.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+/// #![feature(async_await, await_macro, underscore_imports, pin)]
+///
+/// use std::thread;
+/// use std::boxed::PinBox;
+///
+/// use futures::{
+///     channel::oneshot,
+///     future::FutureExt as _,
+///     future::TryFutureExt as _,
+///     executor::spawn,
```
Comment (MajorBreakfast) [inline futures-util/src/compat/tokio.rs]: spawn is now macro. See my [blog post PR](https://github.com/rust-lang-nursery/futures-rs/pull/1165) for a quick intro

File: futures-util/src/compat/tokio.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+///
+/// # Examples
+///
+/// ```ignore
+/// #![feature(async_await, await_macro, underscore_imports, pin)]
+///
+/// use std::thread;
+/// use std::boxed::PinBox;
+///
+/// use futures::{
```
Comment (MajorBreakfast) [inline futures-util/src/compat/tokio.rs]: This style is inconsistent with our other doc examples

File: futures-util/src/compat/tokio.rs (hunk 1 above)
Comment (Nemo157) [reply futures-util/src/compat/tokio.rs]: Nope, I just try and use minor unstable features to give them a good testing. I’ll remove it.

File: futures-util/src/compat/tokio.rs (hunk 4 above)
Comment (Nemo157) [reply futures-util/src/compat/tokio.rs]: Changed

File: futures-util/src/compat/tokio.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+impl Executor for TokioDefaultExecutor {
+    fn spawn_obj(
+        &mut self,
+        task: FutureObj<'static, ()>,
+    ) -> Result<(), SpawnObjError> {
+        let fut = Box::new(task.unit_error().compat(*self));
+        DefaultExecutor::current().spawn(fut).map_err(|err| {
+            panic!(
+                "tokio failed to spawn and doesn't return the future: {:?}",
```
Comment (Nemo157) [inline futures-util/src/compat/tokio.rs]: Here's another argument for not returning the future when spawning fails, I can't see any way to do anything other than panic here.

**Comment by Thomasdezeeuw** [comment]
Shouldn't this be part of the tokio project, rather then futures project?

**Comment by Nemo157** [comment]
That's definitely a good question. I'm mostly concerned about getting something up and running ASAP so that we can start testing futures 0.3 in a wider context.

Maybe we can add this here for now, with a note that it will likely not make it to the 0.3 release, and look into adding a compatibility feature into tokio instead for the release?

**Comment by Thomasdezeeuw** [comment]
Ok, looks good to me.

File: futures-util/src/compat/tokio.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+/// deprecated before or soon after the initial 0.3 release and moved to a
+/// feature in `tokio` instead.
+///
+/// # Examples
+///
+/// ```ignore
+/// #![feature(async_await, await_macro, futures_api, pin, use_extern_macros)]
+///
+/// use std::thread;
+/// use std::boxed::PinBox;
```
Comment (MajorBreakfast) [inline futures-util/src/compat/tokio.rs]: Can you put these two after `use futures::spawn`? (single import block)

File: futures-util/src/compat/tokio.rs (hunk 5 above)
Comment (MajorBreakfast) [reply futures-util/src/compat/tokio.rs]: I think that's fine

Comment (MajorBreakfast) [reply futures-util/src/compat/tokio.rs]: We can always change it if find a better solution 🙂

File: futures-util/src/compat/tokio.rs (hunk 6 above)
Comment (MajorBreakfast) [reply futures-util/src/compat/tokio.rs]: Ah never mind. I'll change it myself 🙂

**Comment by MajorBreakfast** [comment]
I agree that we should get things going ASAP and then move the stuff around until we're happy. Maybe we'll end up putting it into its own crate. Who knows :)

**Comment by Nemo157** [comment]
(What is up with the CI jobs hanging so often?)

**Pull request merged**